server/block/vine.go

```diff
@@ -0,0 +1,285 @@
+func (Vines) EntityInside(_ cube.Pos, _ *world.World, e world.Entity) {
+	if fallEntity, ok := e.(fallDistanceEntity); ok {
+		fallEntity.ResetFallDistance()
+	}
+}
+
+// SetAttachment sets an attachment on the given cube.Direction.
+func (v Vines) SetAttachment(direction cube.Direction, attached bool) Vines {
+	switch direction {
+	case cube.North:
+		v.NorthDirection = attached
+		return v
+	case cube.East:
+		v.EastDirection = attached
+		return v
+	case cube.South:
+		v.SouthDirection = attached
+		return v
+	case cube.West:
+		v.WestDirection = attached
+		return v
+	}
+	panic("should never happen")
+}
+
+// Attachment returns the attachment of the vines at the given direction.
+func (v Vines) Attachment(direction cube.Direction) bool {
+	switch direction {
+	case cube.North:
+		return v.NorthDirection
+	case cube.East:
+		return v.EastDirection
+	case cube.South:
+		return v.SouthDirection
+	case cube.West:
+		return v.WestDirection
+	}
+	panic("should never happen")
+}
+
+// Attachments returns all attachments of the vines.
+func (v Vines) Attachments() (attachments []cube.Direction) {
+	for _, d := range cube.Directions() {
+		if v.Attachment(d) {
+			attachments = append(attachments, d)
+		}
+	}
+	return
+}
+
+// UseOnBlock ...
+func (v Vines) UseOnBlock(pos cube.Pos, face cube.Face, _ mgl64.Vec3, w *world.World, user item.User, ctx *item.UseContext) bool {
+	if _, ok := w.Block(pos).Model().(model.Solid); !ok || face.Axis() == cube.Y {
+		return false
+	}
+	pos, face, used := firstReplaceable(w, pos, face, v)
+	if !used {
+		return false
+	}
+	//noinspection GoAssignmentToReceiver
+	v = v.SetAttachment(face.Direction().Opposite(), true)
+
+	place(w, pos, v, user, ctx)
+	return placed(ctx)
+}
+
+// NeighbourUpdateTick ...
+func (v Vines) NeighbourUpdateTick(pos, _ cube.Pos, w *world.World) {
+	above, updated := w.Block(pos.Side(cube.FaceUp)), false
+	for _, d := range v.Attachments() {
+		if _, ok := w.Block(pos.Side(d.Face())).Model().(model.Solid); !ok {
+			if o, ok := above.(Vines); !ok || ok && !o.Attachment(d) {
+				//noinspection GoAssignmentToReceiver
+				v = v.SetAttachment(d, false)
+				updated = true
+			}
+		}
+	}
+	if !updated {
+		return
+	}
+	if _, ok := above.Model().(model.Solid); !ok && len(v.Attachments()) == 0 {
+		w.SetBlock(pos, nil, nil)
+		return
+	}
+	w.SetBlock(pos, v, nil)
+}
+
+// RandomTick ...
+func (v Vines) RandomTick(pos cube.Pos, w *world.World, r *rand.Rand) {
+	if r.Float64() > 0.25 {
+		// Vines have a 25% chance of spreading.
+		return
+	}
+
+	face := cube.Face(r.Intn(len(cube.Faces())))
+	selectedPos := pos.Side(face)
+	if face == cube.FaceUp || face == cube.FaceDown {
+		if selectedPos.OutOfBounds(w.Range()) {
+			// Vines can't spread outside world bounds.
+			return
+		}
+
+		_, air := w.Block(selectedPos).(Air)
+		newVines, vines := w.Block(selectedPos).(Vines)
+		if face == cube.FaceUp {
+			if !air {
+				// Vines can't spread upwards unless there is air.
+				return
+			}
+
```

Comment:
According to the wiki, there can not be more than 4 vines in a 9x9x3 area to spread upwards